lib/prelude/src/lib.rs

```diff
@@ -74,7 +87,119 @@ pub type PhantomData9<T1,T2,T3,T4,T5,T6,T7,T8,T9> = PhantomData2<PhantomData8<T1
 /// ```
 pub fn with<T, F: FnOnce(T) -> Out, Out>(t: T, f: F) -> Out { f(t) }
 
+pub trait OptionOps {
+    type Item;
+    fn for_each<U, F: FnOnce(Self::Item) -> U>(self, f: F);
+}
+
+impl<T> OptionOps for Option<T> {
+    type Item = T;
+    fn for_each<U, F: FnOnce(Self::Item) -> U>(self, f: F) {
+        match self {
+            Some(x) => { f(x); }
+            None    => {}
+        }
+    }
+}
+
+
+
+
+// ===================
+// === WithPhantom ===
+// ===================
+
+/// A wrapper adding a phantom type to a structure.
+#[derive(Derivative)]
+#[derive(Shrinkwrap)]
+#[shrinkwrap(mutable)]
+#[derivative(Clone   (bound="T:Clone"))]
+#[derivative(Default (bound="T:Default"))]
+#[derivative(Debug   (bound="T:Debug"))]
+pub struct WithPhantom<T, P=()> {
+    #[shrinkwrap(main_field)]
+    pub without_phantom: T,
+    phantom: PhantomData<P>
+}
+
+impl<T, P> WithPhantom<T, P> {
+    pub fn new(without_phantom: T) -> Self {
+        let phantom = PhantomData;
+        Self { without_phantom, phantom }
+    }
+}
```

Comment:
Do we _really_ need this? why not just add phantom field to a structure?

Reply:
This allows just adding phantom field on top of any structure in a nice algebraic fashion. I'm not sure it is so usable as I thought it was. It would be definitely usable in Haskell, here I need more use cases to judge it correctly.